### PR TITLE
MLFlow Integration

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do?

This PR picks up and continues the excellent community contribution from #1542 by @therealnaveenkamal to add comprehensive MLFlow support to Megatron Bridge.

## Features

- **MLFlow Logger Integration**: Add MLFlow logger support in GlobalState with configurable experiment, run name, tracking URI, and tags
- **Metrics Logging**: Log training metrics (losses, learning rate, batch size, throughput, timers, memory, runtime, norms, energy) to MLFlow
- **Checkpoint Artifacts**: Log checkpoint artifacts to MLFlow with iteration-based artifact paths
- **Configuration**: Add MLFlow configuration options to LoggerConfig:
  - `mlflow_experiment`: The MLFlow experiment name
  - `mlflow_run_name`: The MLFlow run name
  - `mlflow_tracking_uri`: Optional MLFlow tracking URI
  - `mlflow_tags`: Optional tags to apply to the MLFlow run
- **Validation**: Add validation in `LoggerConfig.finalize()` to check MLFlow availability
- **Code Refactoring**: Move `safe_serialize` to `log_utils.py` for reuse across WandB and MLFlow
- **Tests**: Add comprehensive unit tests for MLFlow utilities
- **Documentation**: Add MLFlow logging configuration documentation

## Usage

```python
from megatron.bridge.training.config import LoggerConfig

cfg.logger = LoggerConfig(
    tensorboard_dir="./runs/tensorboard",
    mlflow_experiment="my_megatron_experiment",
    mlflow_run_name="llama32_1b_pretrain_run",
    mlflow_tracking_uri="http://mlflow:5000",  # optional
    mlflow_tags={                              # optional
        "project": "llama32",
        "phase": "pretrain",
    },
)
```

## Attribution

This PR is based on the community contribution in #1542 by @therealnaveenkamal. Thank you for the excellent work!

Supersedes #1542

---

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - MLFlow is added as a required dependency in pyproject.toml

cc @therealnaveenkamal @Phlip79

## Summary by CodeRabbit

* **New Features**
  * Added MLFlow integration for comprehensive training logging, supporting metrics, checkpoints, and artifacts alongside WandB and TensorBoard.
  * Configure MLFlow experiment tracking with customizable run names, tags, and tracking URIs.

* **Documentation**
  * Added MLFlow logging setup and configuration guide with installation and enablement instructions.

* **Chores**
  * Added mlflow as a runtime dependency.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by therealnaveenkamal** [comment]
Thanks for continuing this work, @yaoyu-33
